server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDataStreamsService.java

```diff
@@ -210,6 +210,9 @@ static ClusterState updateDataLifecycle(
             var dataStream = validateDataStream(metadata, dataStreamName);
             builder.put(dataStream.copy().setLifecycle(lifecycle).build());
         }
+        if (lifecycle != null) {
+            lifecycle.addWarningHeaderIfDataRetentionNotEffective(DataStreamGlobalRetention.getFromClusterState(currentState));
```

Comment:
Do we not want to use `globalRetentionResolver.resolve(state)` from https://github.com/elastic/elasticsearch/pull/107741/ rather than `DataStreamGlobalRetention.getFromClusterState(currentState)` directly?

Reply:
Yes, you are right. Fixed and added a test to catch it! I will fix the others